### PR TITLE
[E2E refactor - 5] refactor ruler e2e tests

### PR DESCRIPTION
**Pull request by pintohutch**
I decided to break out the kind E2E test refactor PR https://github.com/GoogleCloudPlatform/prometheus-engine/pull/738 into smaller, digestible PRs for reviewing.

Note: because of the nature of the change, presubmits (i.e. Github Actions) may fail until the final PR is merged.

This is the fifth one, where we refactor and simplify rule-evaluator e2e tests.

### CODE REVIEW

File: e2e/ruler_test.go [hunk 1]
```diff
@@ -25,174 +25,212 @@ import (
...
+			t.Fatalf("get operatorconfig: %s", err)
+		}
+
+		// Configure
+		config.Rules.ExternalLabels = map[string]string{
+			"external_key": "external_val"}
+		config.Rules.GeneratorURL = "http://example.com/"
+
+		_, err = opClient.MonitoringV1().OperatorConfigs(operator.DefaultPublicNamespace).Update(ctx, config, metav1.UpdateOptions{})
+		if err != nil {
```
Comment (bwplotka) [inline e2e/ruler_test.go:110]:
```suggestion
		if _, err := opClient.MonitoringV1().OperatorConfigs(operator.DefaultPublicNamespace).Update(ctx, config, metav1.UpdateOptions{}); err != nil {
```

File: e2e/ruler_test.go [hunk 2]
```diff
@@ -25,174 +25,212 @@ import (
...
-							},
-							KeySecret: keySecret,
-						},
-					},
-				},
-			},
-		},
-	})
-}
+		_, err = kubeClient.CoreV1().Secrets(operator.DefaultPublicNamespace).Create(ctx, tlsPair, metav1.CreateOptions{})
```
Comment (bwplotka) [inline e2e/ruler_test.go:179]: all of it has opportunity for simplification (if _, err := ... ; err != nil ), will stop commenting on this one (:

File: e2e/ruler_test.go [hunk 3]
```diff
@@ -25,174 +25,212 @@ import (
...
+		want := map[string][]byte{
+			fmt.Sprintf("secret_%s_alertmanager-tls_cert", operator.DefaultPublicNamespace):            cert,
+			fmt.Sprintf("secret_%s_alertmanager-tls_key", operator.DefaultPublicNamespace):             key,
+			fmt.Sprintf("secret_%s_alertmanager-authorization_token", operator.DefaultPublicNamespace): []byte("auth-bearer-password"),
+		}
+		err = wait.PollUntilContextCancel(ctx, pollDuration, false, func(ctx context.Context) (bool, error) {
+			secret, err := kubeClient.CoreV1().Secrets(operator.DefaultOperatorNamespace).Get(ctx, operator.RulesSecretName, metav1.GetOptions{})
 			if apierrors.IsNotFound(err) {
 				return false, nil
+			} else if err != nil {
```
Comment (bwplotka) [inline e2e/ruler_test.go:201]: ditto (else not needed)

File: e2e/ruler_test.go [hunk 4]
```diff
@@ -25,174 +25,212 @@ import (
...
+			}
+			return true, nil
+		})
+		if err != nil {
+			t.Fatalf("failed waiting for generated rule-evaluator deployment: %s", err)
+		}
+	}
+}
+
+func testRuleEvaluatorConfigured(ctx context.Context, t *testing.T, kubeClient kubernetes.Interface, opClient versioned.Interface) func(*testing.T) {
```
Comment (bwplotka) [inline e2e/ruler_test.go:95]: Hm, the name `testRuleEvaluatorConfigured` and `testRuleEvaluatorConfiguration` are a bit too close, it's hard to understand from function name what's for what. This also indicated the problem of the bit duplicated purpose of those. Can we have one where we ensure OperatorConfig has a custom change and we check rule-eval Prometheus like config?

**Review by bwplotka — APPROVED** [review]
Nice, LGTM mod some suggestions

File: e2e/ruler_test.go (hunk 4 above)
Comment (bwplotka) [reply e2e/ruler_test.go:95]: The same applies to all other tests in other PRs. A bit confusing 🤔

File: e2e/ruler_test.go [hunk 5]
```diff
@@ -25,174 +25,212 @@ import (
...
-		t.Run("check rule metrics", tctx.subtest(testValidateRuleEvaluationMetrics))
+		t.Run("rules-gcm", testValidateRuleEvaluationMetrics(ctx, t))
+	}
+}
+
+func testRuleEvaluatorDeployed(ctx context.Context, t *testing.T, kubeClient kubernetes.Interface) func(*testing.T) {
+	return func(t *testing.T) {
+		t.Log("checking rule-evaluator is running")
+
+		err := wait.PollUntilContextCancel(ctx, pollDuration, false, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (pintohutch) [reply e2e/ruler_test.go:68]: Yea - I was actually a big proponent of that pattern, but it wasn't convention in the repo so I kept with what was here. I would rather have that change in a dedicated commit, where we agree on the convention as a team before doing it here.

File: e2e/ruler_test.go [hunk 6]
```diff
@@ -25,174 +25,212 @@ import (
...
+			deploy, err := kubeClient.AppsV1().Deployments(operator.DefaultOperatorNamespace).Get(ctx, operator.NameRuleEvaluator, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			} else if err != nil {
```
[the remark this answers is not in the document]
Comment (pintohutch) [reply e2e/ruler_test.go:72]: Well we need to check if `err` is not nil after we've validated it's not a "not found" error. If so, then something else is wrong.

File: e2e/ruler_test.go (hunk 1 above)
Comment (pintohutch) [reply e2e/ruler_test.go:110]: Yea - similar to my comment above.

File: e2e/ruler_test.go [hunk 7]
```diff
@@ -25,174 +25,212 @@ import (
...
+		config.Rules.ExternalLabels = map[string]string{
+			"external_key": "external_val"}
+		config.Rules.GeneratorURL = "http://example.com/"
+
+		_, err = opClient.MonitoringV1().OperatorConfigs(operator.DefaultPublicNamespace).Update(ctx, config, metav1.UpdateOptions{})
+		if err != nil {
+			t.Fatalf("update operatorconfig: %s", err)
+		}
+		// Keep checking the state of the collectors until they're running.
+		err = wait.PollUntilContextCancel(ctx, pollDuration, false, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (pintohutch) [reply e2e/ruler_test.go:114]: same with my comment above

File: e2e/ruler_test.go [hunk 8]
```diff
@@ -25,174 +25,212 @@ import (
...
+			deploy, err := kubeClient.AppsV1().Deployments(operator.DefaultOperatorNamespace).Get(ctx, operator.NameRuleEvaluator, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			} else if err != nil {
```
[the remark this answers is not in the document]
Comment (pintohutch) [reply e2e/ruler_test.go:118]: same with my comment above

File: e2e/ruler_test.go (hunk 2 above)
Comment (pintohutch) [reply e2e/ruler_test.go:179]: sgtm :)

File: e2e/ruler_test.go (hunk 4 above)
Comment (pintohutch) [reply e2e/ruler_test.go:95]: Agreed. Changed the name along with the new pattern from collector_test.go to make this more clear.

**Pull request merged**